src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.Utf8JsonReader.cs

```diff
@@ -0,0 +1,323 @@
+        ///   </para>
+        ///
+        ///   <para>
+        ///     This method makes a copy of the data the reader acted on, so there is no caller
+        ///     requirement to maintain data integrity beyond the return of this method.
+        ///   </para>
+        /// </remarks>
+        /// <remarks>
+        /// The <see cref="JsonReaderOptions"/> used to create the instance of the <see cref="Utf8JsonReader"/> take precedence over the <see cref="JsonSerializerOptions"/> when they conflict.
+        /// Hence, <see cref="JsonReaderOptions.AllowTrailingCommas"/>, <see cref="JsonReaderOptions.MaxDepth"/>, <see cref="JsonReaderOptions.CommentHandling"/> are used while reading.
+        /// </remarks>
+        public static object ReadValue(ref Utf8JsonReader reader, Type returnType, JsonSerializerOptions options = null)
+        {
+            if (returnType == null)
+                throw new ArgumentNullException(nameof(returnType));
+
+            return ReadValueCore(ref reader, returnType, options);
+        }
+
+        private static object ReadValueCore(ref Utf8JsonReader reader, Type returnType, JsonSerializerOptions options)
+        {
+            if (options == null)
+            {
+                options = JsonSerializerOptions.s_defaultOptions;
+            }
+
+            ReadStack readStack = default;
+            readStack.Current.Initialize(returnType, options);
+
+            ReadValueCore(options, ref reader, ref readStack);
+
+            return readStack.Current.ReturnValue;
+        }
+
+        private static void CheckSupportedOptions(JsonReaderOptions readerOptions, string paramName)
+        {
+            if (readerOptions.CommentHandling == JsonCommentHandling.Allow)
+            {
+                throw new ArgumentException(SR.JsonSerializerDoesNotSupportComments, paramName);
+            }
+        }
+
+        private static void ReadValueCore(JsonSerializerOptions options, ref Utf8JsonReader reader, ref ReadStack readStack)
+        {
+            JsonReaderState state = reader.CurrentState;
+            CheckSupportedOptions(state.Options, nameof(reader));
+
+            // Value copy to overwrite the ref on an exception and undo the destructive reads.
+            Utf8JsonReader restore = reader;
+
+            ReadOnlySpan<byte> valueSpan = default;
+            ReadOnlySequence<byte> valueSequence = default;
+
+            try
+            {
+                switch (reader.TokenType)
+                {
+                    // A new reader was created and has never been read,
+                    // so we need to move to the first token.
+                    // (or a reader has terminated and we're about to throw)
+                    case JsonTokenType.None:
+                    // Using a reader loop the caller has identified a property they wish to
+                    // hydrate into a JsonDocument. Move to the value first.
+                    case JsonTokenType.PropertyName:
```

Comment:
Hmmmm, the initial read makes sense for a new reader. However I'm not sure a read should happen automatically when you're at a property.

What is an example of it being used?

Reply:
This is consistent with what the `JsonDocument` does.

Also, if you have {"foo":123} and the reader is at property name, it makes sense to deserialize the 123 into an int, since that is the value, no?